Clang-Tidy fixes for findings reported by Zeek in generated code

timwoj opened the pull request:
This PR fixes some issues reported by clang-tidy when running against the generated code for Zeek.

1. The `switch` statements generated by Binpac almost always have cloned branches. These are mostly single break statements. Instead of trying to fix the code gen, just wrap those blocks in `NOLINT` statements to prevent clang-tidy on Zeek builds from spewing a ton of warnings for these.

2. Changes `bool` from being a number underneath to being an actual Boolean type that can be mapped to true/false in c++. This fixes some findings where were having to cast true/false to integers.

3. Adds a `nullptr` keyword for allowing comparisons in Binpac code that would get converted into integer comparisons in the generated code.

timwoj committed:
Wrap generated switch statements in NOLINTs for bugprone-branch-clone

Binpac generates a lot of switch statements with repeated blocks in
them (typically empty blocks). Running clang-tidy on the generated code
with bugprone-branch-clone generates a lot of warnings. Instead of
doing a ton of analysis in binpac to avoid generating the duplicates,
just mark any switch generated with an annotation to avoid reporting
them.

timwoj committed:
Add missing .cmake-format.json, reformat with pre-commit

bbannier commented:
WDYT about instead disabling these "globally" for generated files, i.e., instead of annotating many, many instances in the generated code putting the `NOLINTBEGIN`/`NOLINTEND` around whole generated files? That would make it easier to remove them if we'd ever want to actually inspect the clang-tidy output for these generated files.

That said, I think a better solution for https://github.com/zeek/zeek/pull/4404 might be to just disable linting for these files in Zeek. That would avoid us having to play catch up here as we enable more lints there. One way to do that would be to attach a source property for each generated file, e.g., in `zeek/cmake/BinPAC.cmake`

```patch
diff --git a/BinPAC.cmake b/BinPAC.cmake
index e49ed90..a9f3536 100644
--- a/BinPAC.cmake
+++ b/BinPAC.cmake
@@ -75,6 +75,7 @@ function (binpac_target pacFile)
     # Make paths to generated visible at the caller.
     set(BINPAC_OUTPUT_H "${hdr_file}" PARENT_SCOPE)
     set(BINPAC_OUTPUT_CC "${src_file}" PARENT_SCOPE)
+    set_source_files_properties(${hdr_file} ${src_file} PROPERTIES SKIP_LINTING ON)
 
     # When building Zeek, this target bundles all auto-generated files.
     if (TARGET zeek_autogen_files)
```

timwoj commented:
> That said, I think a better solution for https://github.com/zeek/zeek/pull/4404 might be to just disable linting for these files in Zeek. That would avoid us having to play catch up here as we enable more lints there.

I considered that, but the problem is that there is embedded C++ code in the `.pac` files that gets inserted into the generated files. I'd like all of that to be checked by clang-tidy if possible.

timwoj committed:
Make 'bool' a discrete type and handle it differently when generating code

This fixes clang-tidy modernize-use-bool-literals findings in the generated code

timwoj committed:
Make 'nullptr' a keyword, allow values to be set to it and compared against it

This helps fix 'modernize-use-nullptr' findings in generated code.

timwoj committed:
Remove the unused yapf pre-commit

ckreibich commented on src/pac_exttype.h:
```diff
@@ -21,6 +21,7 @@ class ExternType : public Type {
     string EvalMember(const ID* member_id) const override;
     bool IsNumericType() const override { return ext_type_ == NUMBER; }
     bool IsPointerType() const override { return ext_type_ == POINTER; }
+    bool IsBooleanType() const override { return ext_type_ == BOOLEAN; }
```
If `IsNumericType()` was previously true for Booleans and now is not, I wonder if that could matter. I searched a bit and the only place that caught my eye was [this bit](https://github.com/zeek/binpac/blob/master/src/pac_type.h#L113) in pac_type.h, which I think needs an exclusion also for `IsBooleanType()`.

ckreibich reviewed (approved):
Careful Tim, you'll be our Binpac tech lead before you know it. ;-)

LGTM, feel free to handle that `IsNumericType()`/`IsBooleanType()` point as you see fit.

timwoj committed:
Reorder the fields in pac_type.h for better packing
